nanofi/examples/tail_file.c

```diff
@@ -0,0 +1,223 @@
+ * the License.  You may obtain a copy of the License at
+    *
+    *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+#include "api/nanofi.h"
+#include "core/string_utils.h"
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <limits.h>
+#include <sys/stat.h>
+
+typedef struct flow_file_records {
+    flow_file_record ** records;
+    uint64_t len;
+} flow_file_records;
+
+typedef struct file_offset {
+    int offset;
+} file_offset;
+
+struct flow_file_records * flowfiles = NULL;
+struct file_offset * fileoffset = NULL;
+nifi_instance * instance = NULL;
+standalone_processor * proc = NULL;
+
+void transmit_flow_files(nifi_instance * instance) {
+    if (flowfiles) {
+        for (int i = 0; i < flowfiles->len; ++i) {
+            if (flowfiles->records[i]) {
+                transmit_flowfile(flowfiles->records[i], instance);
+            }
+        }
+    }
+}
+
+void free_flow_file_records() {
+    if (flowfiles != NULL) {
+        for (int i = 0; i < flowfiles->len; ++i) {
+            if (flowfiles->records[i]) {
+                free_flowfile(flowfiles->records[i]);
+            }
+        }
+        free(flowfiles);
+        flowfiles = NULL;
+    }
+}
+
+void set_offset(int offset) {
+    fileoffset->offset = offset;
+}
+
+int get_offset() {
+    return fileoffset->offset;
+}
+
+void free_all_strings(char ** strings, int num_strings) {
+    for (int i = 0; i < num_strings; ++i) {
+        if (strings[i]) free(strings[i]);
+    }
+}
+
+void on_trigger_callback(processor_session * ps, processor_context * ctx) {
+
+    char file_path[50];
+    char delimiter[2];
+
+    if (get_property(ctx, "file_path", file_path, 50) != 0) {
+        return;
+    }
+
+    if (get_property(ctx, "delimiter", delimiter, 2) != 0) {
+        return;
+    }
+
+    if (strlen(delimiter) == 0) {
+        printf("Delimiter not specified or it is empty\n");
+        return;
+    }
+    char delim = '\0';
+    if (strlen(delimiter) > 0) {
+        delim = delimiter[0];
+    }
+
+    if (delim == '\0') {
+        printf("Invalid delimiter \n");
+        return;
+    }
+
+    int curr_offset = get_offset();
+    int max_bytes_read = 4096;
+    char buff[max_bytes_read];
+    memset(buff,'\0', max_bytes_read);
+    FILE * fp = fopen(file_path, "rb");
+    if (!fp) return;
+    fseek(fp, curr_offset, SEEK_SET);
+
+    int bytes_read = fread(buff, 1, max_bytes_read, fp);
+    if (bytes_read > 0) {
+        struct tokens * tks = tokenize_string(buff, delimiter);
+
+        if (!tks || tks->num_strings == 0) return;
+
+        set_offset(get_offset() + tks->total_bytes);
```

Comment:
I think the entire processor will stop working if there is a line longer than max_bytes_read-1 bytes in the file: you only create tokens for entire lines, and seek the file after the end of the last processed line. But if there is no delimiter in max_bytes_read-1 bytes, you will never be able to exit this condition and will try to read this line and fail every time you are triggered.

Reply:
I think there is no harm caching the current line in memory if it has not ended in the current read.
Also, I don't think you should only read once every trigger: you should read and create FlowFile until you hit EOF.